### ISSUE TITLE
adapter: improve notice wording when cluster becomes not ready

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#15692]
### Feature request

@frankmcsherry reports that the current wording of the "cluster became not ready" notice that we added in #15262 is not as clear and actionable as it could be!

```
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
```

Seems like a good moment to make use of the detail/hint fields!

It also seems like the notice is being emitted for replicas of all clusters, not just replicas of the cluster to which the user's session is connected. 

cc @mjibson @chaas @ggnall

**Comment by brennanvincent** [comment]
I am noticing a large amount of `NotReady` spam when dropping replicas. For example, it printed 21 times after dropping a 2xlarge.

**Comment by mjibson** [comment]
> It also seems like the notice is being emitted for replicas of all clusters, not just replicas of the cluster to which the user's session is connected.

Evidence for this? It's supposed to be handled at https://github.com/MaterializeInc/materialize/blob/88a0a20b0bcdcd2f312e2638bbb7096ed8fef1a4/src/adapter/src/session.rs#L346

**Comment by benesch** [comment]
> > It also seems like the notice is being emitted for replicas of all clusters, not just replicas of the cluster to which the user's session is connected.
> 
> Evidence for this? It's supposed to be handled at
> 
> https://github.com/MaterializeInc/materialize/blob/88a0a20b0bcdcd2f312e2638bbb7096ed8fef1a4/src/adapter/src/session.rs#L346

I was looking at two copy/pasted outputs in Slack that show intermixed output about both `mz_system` and `mz_introspection`:
 * https://materializeinc.slack.com/archives/C02FWJ94HME/p1666385319999599
 * https://materializeinc.slack.com/archives/C02FWJ94HME/p1666622997299049

I guess that's not quite a smoking gun, but it sure did seem suspicious!

**Comment by mjibson** [comment]
Ah! Yeah what's going on there? Logs:

```
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
NOTICE:  cluster replica mz_introspection.r1 changed status to: NotReady
NOTICE:  cluster replica mz_introspection.r1 changed status to: NotReady
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
NOTICE:  cluster replica mz_system.r1 changed status to: NotReady
NOTICE:  cluster replica mz_introspection.r1 changed status to: NotReady
```

**Comment by chaas** [comment]
@morsapaes @sjwiesman Do you have thoughts on what wording would be more clear here. There isn't really anything actionable except for contacting support and/or waiting for it to resolve.

Perhaps something like
```
NOTICE: cluster replica mz_system.r1 changed status to: NotReady.
HINT: The cluster replica may be restarting or going offline.
``` 
I don't want to cause too much alarm b/c it could be a "false alarm" in that the status is flapping for some reason, or on startup of a new cluster, but the notice is useful in discovering that replicas are OOMing or in a restart loop which can otherwise present to the user as a slow query.

**Comment by sjwiesman** [comment]
Maybe controversial, but maybe we put this notice behind unsafe until there’s something in mz_catalog users can query to see what’s happening?

**Comment by mjibson** [comment]
```
materialize=> select * from mz_internal.mz_cluster_replica_statuses;
 replica_id | process_id | status |          last_update          
------------+------------+--------+-------------------------------
 11         |          0 | ready  | 2022-11-03 19:34:05.16131+00
 12         |          0 | ready  | 2022-11-03 19:34:04.297245+00
 13         |          0 | ready  | 2022-11-03 19:34:04.41506+00
 19         |          0 | ready  | 2022-11-03 19:34:04.590684+00
 21         |          0 | ready  | 2022-11-04 00:36:04.164133+00
 24         |          0 | ready  | 2022-11-03 19:34:04.973068+00
 25         |          0 | ready  | 2022-11-03 19:34:05.042935+00
 26         |          0 | ready  | 2022-11-03 19:34:05.108107+00
 27         |          0 | ready  | 2022-11-03 19:34:04.52519+00
 28         |          0 | ready  | 2022-11-03 19:34:04.722281+00
(10 rows)
```

Is a thing, but it's in mz_internal so we caution it's unstable, so maybe this isn't what you want either?

**Comment by morsapaes** [comment]
If this isn't always a sign that something has gone haywire, maybe we could consider documenting it in `Troubleshooting` as something to check for when faced with slow queries (+ where to look) vs. emitting a possibly alarming notice? This seems to have caused a good deal of confusion with Previews users.

I'd +1 on Seth's suggestion, though from [Nikhil's comment](https://github.com/MaterializeInc/materialize/pull/15262#issuecomment-1272581974) it sounds like this might be removed altogether once the observability work is complete?

**Comment by mjibson** [comment]
I understand this message to indeed mean things have gone haywire. I personally don't care whether or not we have this notice. It was added because we (I?) thought it would help people be more able to track down invisibly slow query problems. But if it's causing more confusion then it's not meeting it's goal and could be rethought.

**Comment by umanwizard** [comment]
+1 to getting rid of it. I personally think it's confusing. I get a random number of copies of this notice at random times, including when the replica is actually ready and responding to queries fine. Maybe it means something useful, but not that I've been able to discover.

**Comment by mjibson** [comment]
How do you know the replica is ready if you are getting these? If that's the case then this is a bug (and the notice is working as intended).

**Comment by umanwizard** [comment]
It's possible what I'm seeing is _old_ messages from a replica I already dropped; not sure.

**Comment by umanwizard** [comment]
There seem to be a few separate issues:

(1) the number of messages I get per replica is apparently random
(2) the messages show for a dropped replica
(3) they are held for arbitrarily long, so I might get a few of them way later, when I send my next SQL command

**Comment by mjibson** [comment]
1) should be fixed with some de-dup [code](https://github.com/MaterializeInc/materialize/pull/15714/commits/20cfd06d32c3232f295fc054e744d4a42fe5e271) that got added in 0.30.0
2 and 3 are expected behavior. In general I do not expect users to be dropping replicas much, so I think this behavior is fine for them.

**Comment by umanwizard** [comment]
I am still seeing duplicate notices on cloud, which post-dates 0.30.0.

**Comment by umanwizard** [comment]
Another issue, closely related to (3):

(4) If the replicas become ready in between when the messages were queued and when you send your next SQL statement, the messages are apparently not cleared, and still sent, so you will be notified that things are not ready that in fact are.

```
materialize=> create cluster btv1 replicas (r1 (size '2xlarge'), r2 (size 'xlarge'), r3 (size '2xsmall'));
CREATE CLUSTER

(Here I waited, long enough for them all to become ready)

materialize=> select 1;
NOTICE:  cluster replica btv1.r2 changed status to: NotReady
NOTICE:  cluster replica btv1.r1 changed status to: NotReady
NOTICE:  cluster replica btv1.r3 changed status to: NotReady
NOTICE:  cluster replica btv1.r2 changed status to: NotReady
NOTICE:  cluster replica btv1.r1 changed status to: NotReady
NOTICE:  cluster replica btv1.r3 changed status to: NotReady
NOTICE:  cluster replica btv1.r1 changed status to: NotReady
NOTICE:  cluster replica btv1.r3 changed status to: NotReady
NOTICE:  cluster replica btv1.r1 changed status to: NotReady
NOTICE:  cluster replica btv1.r1 changed status to: NotReady
NOTICE:  cluster replica btv1.r2 changed status to: NotReady
NOTICE:  cluster replica btv1.r1 changed status to: NotReady
 ?column? 
----------
        1
(1 row)

materialize=> select * from mz_internal.mz_cluster_replica_statuses;
 replica_id | process_id | status |          last_update          
------------+------------+--------+-------------------------------
 1          |          0 | ready  | 2022-11-07 15:37:29.73979+00
 2          |          0 | ready  | 2022-11-07 15:37:28.71426+00
 3          |          0 | ready  | 2022-11-07 15:37:28.766506+00
 22         |          0 | ready  | 2022-11-07 18:32:04.718914+00
 22         |          1 | ready  | 2022-11-07 18:32:04.65722+00
 23         |          0 | ready  | 2022-11-07 18:32:05.331088+00
 24         |          0 | ready  | 2022-11-07 18:32:05.279673+00
(7 rows)

materialize=> 
```

**Comment by mjibson** [comment]
Because the replica changed each time, receiving interleaved messages like this is also working as intended (it only de-dups from the previous message).

4) Yes, also working as intended. Unsending pending notices isn't a thing we care about.

IMO: you are seeing all of this because you are poking at stuff in ways users will not, and it's ok. You can disable notices yourself with `SET client_min_messages TO WARNING;`, which can be added to a `~/.psqlrc` file to auto execute.

I understood our problem to be: when users experience slow queries, we would like them to have an idea why, and this notice is that.
